mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
     public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        save(mailbox, addUidAndModseq(message, mailboxId))
-            .thenEmpty(indexTableHandler.updateIndexOnAdd(message, mailboxId))
-            .block();
+        block(addUidAndModseq(message, mailboxId)
+            .flatMap(Throwing.function(messageWithUidAndModSeq -> save(mailbox, messageWithUidAndModSeq)))
+            .thenEmpty(Mono.defer(() -> indexTableHandler.updateIndexOnAdd(message, mailboxId))));
```

Comment:
Why did you add this `defer`?

Reply:
Overwize mutation on message didn't happen.

Reply:
I don't understand, sorry. Could you elaborate?

Reply:
You are right, mutability is ugly, let's hide it.